File: javascript/stack/stack.js
'use strict';

class Node {
  constructor(data) {
    this.data = data;
  }
}

class Stack {

  constructor(size) {
    this.size = size;
    this.stack = [];
    this.top = -1;
    this.create(size);
  }
  create(size) {
    for(let i=0;i<size;i++) {
      this.stack.push(null);
    }
  }

  push(data) {
    if (!this.isFull()) {
      console.log(`PUSH - top -> ${this.top}, data -> ${data}`)
      this.top++;
      this.stack[this.top] = data;
    } else {
      console.log('PUSH - Cannot add ' + data + ' since stack is full');
    }
  }

  pop() {
    if(this.isEmpty()) {
      console.log('POP - Cannot remove - Nothing to pop since stack is empty');
    } else {
      console.log(`POP - top -> ${this.top}`);
      this.stack[this.top] = null;
      this.top--;
    }
  }

  peek() {
    if(this.isEmpty()) {
      console.log('Nothing to peek since stack is empty');
    } else {
      console.log('The top of the stack is top[' + this.stack[this.top] + ']');
    }
  }

  isEmpty() {
    if(this.top >= 0) {
      console.log("The stack is not empty");
      return false;
    } else {
      console.log("The stack is empty");
      return true;
    }
  }

  isFull() {
    if (this.top < this.size-1) {
      console.log("The stack is not full");
      return false;
    } else {
      console.log("The stack is full");
      return true;
    }
  }

  size() {
    console.log('The size of the stack is ' + this.size);
  }  

  printAll() {
    console.log('=======');
    console.log('STACK');
    let output = this.stack.reduceRight((acc, item) => { return acc + '[' + item + ']\n' }, []);
    console.log(output);
    console.log('=======');
  }
}


let stack = new Stack(5);
stack.push(1);
stack.printAll();
stack.push(2);
stack.printAll();
stack.push(3);
stack.printAll();
stack.pop();
stack.printAll();
stack.pop();
stack.printAll();
stack.pop();
stack.printAll();
stack.pop();
stack.printAll();

File: python/stack.py
class stack:
    def __init__(self):
        self.data = []
    def push(self, item):
        self.data.append(item)
    def pop(self):
        if len(self.data) > 0:
            self.data.pop()
    def top(self):
        if len(self.data) > 0:
            return self.data[-1]
        else:
            return None
    def val(self, index):
        if 0 <= index and index < len(self.data):
            return self.data[index]
        else:
            return None
    def list(self):
        return self.data
    def size(self):
        return len(self.data)
    def clear(self):
        self.data = []

S = stack()
S.push(10)
S.push(20)
S.push(30)
S.push(40)
S.push(50)
print(S.list())
# [10 20 30 40 50]
print(S.val(1))
# 20
print(S.top())
# 50
S.pop()
print(S.top())
# 40
print(S.size())
# 4
S.clear()
print(S.list())
# []
